src/Compilers/Core/CommandLine/BuildProtocol.cs

```diff
@@ -43,14 +45,17 @@ internal class BuildRequest
         public readonly uint ProtocolVersion;
         public readonly RequestLanguage Language;
         public readonly ReadOnlyCollection<Argument> Arguments;
+        public readonly string CompilerHash;
 
         public BuildRequest(uint protocolVersion,
                             RequestLanguage language,
+                            string compilerHash,
                             IEnumerable<Argument> arguments)
         {
```

Comment:
Should add a `Debug.Assert` to verify `compilerHash` is not null or empty. That only happens when we can't detect the hash and in that case we shouldn't be going to the server.